guardian/pkg/asyncutil/async.go

```diff
@@ -0,0 +1,274 @@
+// Resuming ensures the backlogged cmds are then executed again.
+type CommandExecutor[C any, R any] interface {
+	Send(C) <-chan Result[R]
+	ExecutionController
+}
+
+// ExecutionController is an interface to manage the execution of commands of an Executor. It allows you to stop and resume
+// command execution, as well as retrieve a signaler to notify you about shutdown.
+type ExecutionController interface {
+	DrainAndBacklog() Signaler
+	Resume()
+	ShutdownSignaler() Signaler
+}
+
+type executorCoordinator []ExecutionController
+
+func (coordinator executorCoordinator) DrainAndBacklog() Signaler {
+	var signalers []Signaler
+	for _, executor := range coordinator {
+		signalers = append(signalers, executor.DrainAndBacklog())
+	}
+
+	signal := NewSignaler()
+	go func() {
+		defer signal.Send()
+		for _, signaler := range signalers {
+			<-signaler.Receive()
+		}
+	}()
+
+	return signal
+}
+
+func (coordinator executorCoordinator) Resume() {
+	for _, executor := range coordinator {
+		executor.Resume()
+	}
+}
+
+func (coordinator executorCoordinator) ShutdownSignaler() Signaler {
+	var signalers []Signaler
+	for _, executor := range coordinator {
+		signalers = append(signalers, executor.ShutdownSignaler())
+	}
+
+	signal := NewSignaler()
+	go func() {
+		defer signal.Send()
+		for _, signaler := range signalers {
+			<-signaler.Receive()
+		}
+	}()
+
+	return signal
+}
+
+type commandExecutor[C any, R any] struct {
+	command             func(context.Context, C) (R, error)
+	drainAndBacklogSig  chan Signaler
+	resumeBackloggedSig Signaler
+	cmdChan             chan Command[C, R]
+	// backlogChan contains all the commands that failed with EOF, waiting to be retried.
+	backlogChan chan Command[C, R]
+	// inflightCmds keeps track of the number of commands that are currently being executed.
+	inflightCmds        sync.WaitGroup
+	executeSig          Signaler
+	shutdownCompleteSig Signaler
+	errBuff             ErrorBuffer
+
+	backLogCommands bool
+	backlog         []Command[C, R]
+}
+
+func NewCommandCoordinator(executors ...ExecutionController) ExecutionController {
+	var coordinator executorCoordinator
+	for _, executor := range executors {
+		coordinator = append(coordinator, executor)
+	}
+	return coordinator
+}
+
+// NewCommandExecutor creates a new CommandExecutor implementation. It calls the given function f with the command given
+// to Send. Any errors from the function are sent over the errBuff. If an EOF is sent over the error buff, the caller
+// must pause the executor, restart / fix whatever processes need restarting or fixing, then resume execution (using the
+// PauseExecution and ResumeExecution functions). ResumeExecution re runs the commands that failed with EOF.
+func NewCommandExecutor[C any, R any](ctx context.Context, errBuff ErrorBuffer, f func(context.Context, C) (R, error)) CommandExecutor[C, R] {
+	hdlr := &commandExecutor[C, R]{
```

Comment:
typo: mind using executor or commandExecutor  instead of hdlr ?

Reply:
Not at all this was from a previous iteration.